lightningd/opening_control.c

```diff
@@ -820,65 +821,87 @@ static void openchannel_hook_cb(struct openchannel_hook_payload *payload STEALS,
 
-	/* If we had a hook, check what it says */
-	if (buffer) {
-		const jsmntok_t *t = json_get_member(buffer, toks, "result");
-		if (!t)
-			fatal("Plugin returned an invalid response to the"
-			      " openchannel hook: %.*s",
-			      toks[0].end - toks[0].start,
-			      buffer + toks[0].start);
-
-		if (json_tok_streq(buffer, t, "reject")) {
-			t = json_get_member(buffer, toks, "error_message");
-			if (t)
-				errmsg = json_strdup(tmpctx, buffer, t);
-			else
-				errmsg = "";
-			log_debug(openingd->ld->log,
-				  "openchannel_hook_cb says '%s'",
-				  errmsg);
-			our_upfront_shutdown_script = NULL;
-		} else if (!json_tok_streq(buffer, t, "continue"))
-			fatal("Plugin returned an invalid result for the "
-			      "openchannel hook: %.*s",
-			      t->end - t->start, buffer + t->start);
-
-		/* Check for a 'close_to' address passed back */
-		if (!errmsg) {
-			t = json_get_member(buffer, toks, "close_to");
-			if (t) {
-				switch (json_to_address_scriptpubkey(tmpctx, chainparams,
-								     buffer, t,
-								     &our_upfront_shutdown_script)) {
-					case ADDRESS_PARSE_UNRECOGNIZED:
-						fatal("Plugin returned an invalid response to the"
-						      " openchannel.close_to hook: %.*s",
-						      t->end - t->start, buffer + t->start);
-					case ADDRESS_PARSE_WRONG_NETWORK:
-						fatal("Plugin returned invalid response to the"
-						      " openchannel.close_to hook: address %s is"
-						      " not on network %s",
-						      tal_hex(NULL, our_upfront_shutdown_script),
-						      chainparams->network_name);
-					case ADDRESS_PARSE_SUCCESS:
-						errmsg = NULL;
-				}
-			} else
-				our_upfront_shutdown_script = NULL;
-		}
-	} else
-		our_upfront_shutdown_script = NULL;
-
 	subd_send_msg(openingd,
 		      take(towire_openingd_got_offer_reply(NULL, errmsg,
 							  our_upfront_shutdown_script)));
 }
 
-REGISTER_SINGLE_PLUGIN_HOOK(openchannel,
-			    openchannel_hook_cb,
-			    openchannel_hook_serialize,
-			    struct openchannel_hook_payload *);
+static bool
+openchannel_hook_deserialize(struct openchannel_hook_payload *payload,
+			     const char *buffer,
+			     const jsmntok_t *toks)
+{
+	struct subd *openingd = payload->openingd;
+
+	/* already rejected by prior plugin hook in the chain */
+	if (payload->errmsg != NULL)
+		return true;
+
+	if (!toks || !buffer)
+		return true;
+
+	const jsmntok_t *t = json_get_member(buffer, toks, "result");
+	if (!t)
+		fatal("Plugin returned an invalid response to the"
+		      " openchannel hook: %.*s",
+		      toks[0].end - toks[0].start,
+		      buffer + toks[0].start);
+
+	/* reject */
+	if (json_tok_streq(buffer, t, "reject")) {
+		payload->errmsg = "";
+		t = json_get_member(buffer, toks, "error_message");
+		if (t)
+			payload->errmsg = json_strdup(payload, buffer, t);
+		log_debug(openingd->ld->log,
+			  "openchannel_hook rejects and says '%s'",
+			  payload->errmsg);
+		t = json_get_member(buffer, toks, "close_to");
+		if (t)
```

Comment:
Nit: I usually prefer to have separate variables to the tokens and search for all of them initially, so we don't get confused by the same variable name assuming different meanings, and making composite checks simpler to grok.

Reply:
Can be addressed in a cleanup

Reply:
great, looks a lot better that way